go/vault/api/api.go

```diff
@@ -0,0 +1,135 @@
+	ErrNoSuchAction = errors.New(ModuleName, 6, "vault: no such action")
+	// ErrUnsupportedAction is the error returned when an action is not supported.
+	ErrUnsupportedAction = errors.New(ModuleName, 7, "vault: action not supported")
+)
+
+// Backend is a vault implementation.
+type Backend interface {
+	// Vaults returns all of the registered vaults.
+	Vaults(ctx context.Context, height int64) ([]*Vault, error)
+
+	// Vault returns information about the given vault.
+	Vault(ctx context.Context, query *VaultQuery) (*Vault, error)
+
+	// AddressState returns the state information for the given source address.
+	AddressState(ctx context.Context, query *AddressQuery) (*AddressState, error)
+
+	// PendingActions returns the list of pending actions for the given vault.
+	PendingActions(ctx context.Context, query *VaultQuery) ([]*PendingAction, error)
+
+	// StateToGenesis returns the genesis state at specified block height.
+	StateToGenesis(ctx context.Context, height int64) (*Genesis, error)
+
+	// ConsensusParameters returns the vault consensus parameters.
+	ConsensusParameters(ctx context.Context, height int64) (*ConsensusParameters, error)
+
+	// GetEvents returns the events at specified block height.
+	GetEvents(ctx context.Context, height int64) ([]*Event, error)
+
+	// WatchEvents returns a channel that produces a stream of Events.
+	WatchEvents(ctx context.Context) (<-chan *Event, pubsub.ClosableSubscription, error)
+
+	// Cleanup cleans up the backend.
+	Cleanup()
+}
+
+// VaultQuery is a query for data about a given vault.
+type VaultQuery struct {
+	// Height is the query height.
+	Height int64 `json:"height"`
+	// Address is the vault address.
+	Address staking.Address `json:"address"`
+}
+
+// AddressQuery is a query for data about a given address for the given vault.
+type AddressQuery struct {
+	// Height is the query height.
+	Height int64 `json:"height"`
+	// Vault is the vault address.
+	Vault staking.Address `json:"vault"`
+	// Address is the queried address.
+	Address staking.Address `json:"address"`
+}
+
+// Genesis is the initial vault state for use in the genesis block.
+type Genesis struct {
+	// Parameters are the genesis consensus parameters.
+	Parameters ConsensusParameters `json:"params"`
+
+	// Vaults are the vaults.
+	Vaults []*Vault `json:"vaults,omitempty"`
+	// States are the per vault per-address states.
+	States map[staking.Address]map[staking.Address]*AddressState `json:"states,omitempty"`
+	// PendingActions are the per-vault pending actions.
+	PendingActions map[staking.Address][]*PendingAction `json:"pending_actions,omitempty"`
+}
+
+// ConsensusParameters are the vault consensus parameters.
+type ConsensusParameters struct {
+	// Enabled specifies whether the vault service is enabled.
+	Enabled bool `json:"enabled,omitempty"`
+
+	// MaxAuthorityAddresses is the maximum number of addresses that can be configured for each
+	// authority.
+	MaxAuthorityAddresses uint16 `json:"max_authority_addresses,omitempty"`
```

Comment:
Is `uint8` too small?

Reply:
It is not `uint8` but rather `uint16`, that should be enough?

Reply:
Wanted to ask why we don't use `uint8` (same as threshold). Is it because it's too small? Maybe it is, but update action would still need to set at 256 addresses, which is 21*256 = ~5kB.

Reply:
Ah yes, I think `uint8` should be fine. The current configuration is `32` anyway.

Reply:
Changed to `uint8`.